### DISCUSSION TITLE
How to target older SDK than host macOS

### DISCUSSION THREAD

**Discussion by jeroen** [Homebrew/discussions#4808] [Tap maintenance and Homebrew development] [answered]
I would like to build some bottles for `arm64_big_sur` on a M1 machine running `arm64_ventura` with the big-sur SDK installed. 

When I manually compile things I set the following variables to target big-sur:

```
export SDKROOT=/Library/Developer/CommandLineTools/SDKs/MacOSX11.sdk
export MACOSX_DEPLOYMENT_TARGET=11.0
export OS_VER=20
```

Is there a way to trick homebrew to target an other SDK version than the host, such that it will download `arm64_big_sur` bottles for the dependencies, and target this SDK when building bottles?

**Comment by Bo98** [comment, marked as answer]
Not officially, but the idea of how to do this would be to trick `MacOS.full_version` (or `sw_vers`) and then pretty much everything else will work from there. Internally we have a `HOMEBREW_FAKE_EL_CAPITAN`, but that doesn't extend to arbitrary versions.

When doing anything like this though, I recommend enabling cross-compiling in any configure scripts (e.g. `--build=aarch64-apple-darwin22 --host=aarch64-apple-darwin20`). That is because some may do runtime tests when cross-compiling is disabled, which will misconfigure your build for your target.

**Comment by jeroen** [comment]
OK for future reference in case somebody needs this:

I had difficulty getting this to work on ventura, homebrew kept complaining about missing big-sur command line utilities, but it is not possible to install big-sur command line utilities on ventura. Also it is tricky to make homebrew target the correct SDK.

However I found a neat solution:

In ventura, we cannot virtualize big sur, but we can easily virtualize macos 12 (monterey). So we start with a basic monterey vm:

```
tart clone ghcr.io/cirruslabs/macos-monterey-base:latest monterey
tart run monterey
```

Now we download and install `Command_Line_Tools_for_Xcode_13.dmg` from: https://developer.apple.com/xcode/resources/
Note that it has to be exactly v13.0 and not newer. This version can be installed on MacOS 12 (monterey), and if we look in the [xcode support table](https://developer.apple.com/support/xcode/) version 13 of the command line utilities target the `macOS 11.3` SDK.

So we now have a virtual machine that has (only) the big-sur SDK and command line utilities, but still can be virtualized. So the only thing left to do is what Bo98 says above, either shim `sw_vers` to say `11.7.10` or manually hack `HOMEBREW_MACOS_VERSION` in  `brew.sh` and voila, we have a fully working big-sur build environment.

**Reply by ideoforms** [reply]
Fantastic tip on using `tart` as a portable build environment - this is now working perfectly for my situation. Thanks @jeroen!
